logging/impl.go

```diff
@@ -192,89 +208,197 @@ func (imp *impl) formatw(logLevel Level, msg string, keysAndValues ...interface{
 
 func (imp *impl) Debug(args ...interface{}) {
 	if imp.shouldLog(DEBUG) {
-		imp.log(imp.format(DEBUG, args...))
+		imp.log(imp.format(DEBUG, emptyTraceKey, args...))
+	}
+}
+
+func (imp *impl) CDebug(ctx context.Context, args ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for debug, or if there's a trace key.
+	if imp.shouldLog(DEBUG) || dbgName != emptyTraceKey {
+		imp.log(imp.format(DEBUG, dbgName, args...))
 	}
 }
 
 func (imp *impl) Debugf(template string, args ...interface{}) {
 	if imp.shouldLog(DEBUG) {
-		imp.log(imp.formatf(DEBUG, template, args...))
+		imp.log(imp.formatf(DEBUG, emptyTraceKey, template, args...))
+	}
+}
+
+func (imp *impl) CDebugf(ctx context.Context, template string, args ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for debug, or if there's a trace key.
+	if imp.shouldLog(DEBUG) || dbgName != emptyTraceKey {
+		imp.log(imp.formatf(DEBUG, dbgName, template, args...))
 	}
 }
 
 func (imp *impl) Debugw(msg string, keysAndValues ...interface{}) {
 	if imp.shouldLog(DEBUG) {
-		imp.log(imp.formatw(DEBUG, msg, keysAndValues...))
+		imp.log(imp.formatw(DEBUG, emptyTraceKey, msg, keysAndValues...))
+	}
+}
+
+func (imp *impl) CDebugw(ctx context.Context, msg string, keysAndValues ...interface{}) {
+	dbgName := GetName(ctx)
+
+	// We log if the logger is configured for debug, or if there's a trace key.
+	if imp.shouldLog(DEBUG) || dbgName != emptyTraceKey {
```

Comment:
[q] So if we're on info-level logging, we would still log a debug-level log that contained a trace key? That makes sense to me, but just want to verify.

Reply:
Correct. The log level outputted is 1-1 with the log function called.

Reply:
Offline clarification:
If the log "filter" is at INFO, a "debug context" logs at the level of the written log line. E.g:
```
logger.SetLevel(INFO)
logger.Debug("won't output")
logger.CDebug(EnableDebugMode(ctx), "will output at debug")
```
Gives a log line such as:
```
2023-10-30T09:12:09.459Z	DEBUG	impl	logging/impl_test.go:200	will output at debug	{"traceKey":"foobar"}
```